### PR TITLE
Added GC public library

### PR DESCRIPTION
**Pull request by gaius-qi**
Signed-off-by: Gaius <gaius.qi@gmail.com>

## Description
- Added GC public library.

## Related Issue
#266 

## Motivation and Context
Standardize the GC usage of each module.

## Screenshots (if appropriate):
N/A
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist:

- [x] Code compiles correctly.
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: pkg/gc/gc.go [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+	})
+}
+
+func (g gc) run(ctx context.Context, k string, t Task) {
+	done := make(chan struct{})
+
+	go func() {
+		g.logger.Infof("%s GC %s", k, "start")
+		defer close(done)
+		if err := t.RunGC(); err != nil {
```
Comment (yangjun289519474) [inline pkg/gc/gc.go:175]: t.RunGC cann't exit when gc.Stop()

**Review by yangjun289519474 — APPROVED** [review]

**Review by garfield009 — APPROVED** [review]

File: pkg/gc/gc.go [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+type gc struct {
+	tasks    *sync.Map
+	interval time.Duration
+	timeout  time.Duration
+	logger   Logger
+	done     chan bool
+}
+
+// Option is a functional option for configuring the GC
+type Option func(g *gc) *gc
```
Comment (244372610) [inline pkg/gc/gc.go:56]: The pointer variable has been used and should not return a value

File: pkg/gc/gc.go [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+func WithLogger(logger Logger) Option {
+	return func(g *gc) *gc {
+		g.logger = logger
+		return g
+	}
+}
+
+// New returns a new GC instence
+func New(options ...Option) (GC, error) {
+	return NewWithOptions(options...)
```
Comment (244372610) [inline pkg/gc/gc.go:84]: It should be distinguished from the following approach, where a no-argument function is used and the default configuration option is passed in

File: pkg/gc/gc.go [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+// GC is the interface used for release resource
+type GC interface {
+	// Add adds GC task
+	Add(string, Task)
```
Comment (244372610) [inline pkg/gc/gc.go:31]: The Task interface also needs to have a timeout and an execution interval

File: pkg/gc/gc.go [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+	g.tasks.Store(k, t)
+}
+
+func (g gc) Run(k string) error {
+	v, ok := g.tasks.Load(k)
+	if !ok {
+		return errors.New("can not find the task")
+	}
+
+	go g.run(context.Background(), k, v.(Task))
```
Comment (244372610) [inline pkg/gc/gc.go:116]: Whether the context variable should be mentioned outside the method and used to cancel the execution of the task

File: pkg/gc/gc.go [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+	return nil
+}
+
+func (g gc) RunAll() {
+	g.runAll(context.Background())
```
Comment (244372610) [inline pkg/gc/gc.go:121]: same as above

File: pkg/gc/gc.go [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+
+func (g gc) runAll(ctx context.Context) {
+	g.tasks.Range(func(k, v interface{}) bool {
+		go g.run(ctx, k.(string), v.(Task))
+		return true
+	})
+}
+
+func (g gc) run(ctx context.Context, k string, t Task) {
+	done := make(chan struct{})
```
Comment (244372610) [inline pkg/gc/gc.go:170]: Whether you need to confirm that the last GC has ended or cancel the last GC to execute

File: pkg/gc/gc.go (hunk 4 above)
Comment (gaius-qi) [reply pkg/gc/gc.go:31]: There is no such demand now.
https://wiki.c2.com/?PrematureOptimization

File: pkg/gc/gc.go (hunk 5 above)
Comment (gaius-qi) [reply pkg/gc/gc.go:116]: This is a separate module package.

File: pkg/gc/gc.go (hunk 6 above)
Comment (gaius-qi) [reply pkg/gc/gc.go:121]: This is a separate module package.
https://github.com/hanwen/go-fuse/blob/0f728ba15b38579efefc3dc47821882ca18ffea7/benchmark/statfs.go

File: pkg/gc/gc.go (hunk 7 above)
Comment (gaius-qi) [reply pkg/gc/gc.go:170]: The gc function needs to be idempotent.

File: pkg/gc/gc.go (hunk 3 above)
Comment (gaius-qi) [reply pkg/gc/gc.go:81]: https://github.com/kubernetes/kubernetes/blob/cea1d4e20b4a7886d8ff65f34c6d4f95efcb4742/staging/src/k8s.io/sample-controller/pkg/generated/informers/externalversions/factory.go#L87

**Review by yangjun289519474 — APPROVED** [review]

**Review by 244372610 — APPROVED** [review]

**Pull request merged**